flink-runtime/src/test/java/org/apache/flink/runtime/rest/FileUploadHandlerTest.java

```diff
@@ -0,0 +1,483 @@
+			assertEquals(HttpResponseStatus.INTERNAL_SERVER_ERROR.code(), response.code());
+		}
+
+		Request fileRequest = buildFileRequest(mixedHandler.getMessageHeaders().getTargetRestEndpointURL());
+		try (Response response = client.newCall(fileRequest).execute()) {
+			// expected JSON payload is missing
+			assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.code());
+		}
+
+		int mixedId = RANDOM.nextInt();
+		Request mixedRequest = buildMixedRequest(mixedHandler.getMessageHeaders().getTargetRestEndpointURL(), mixedId);
+		try (Response response = client.newCall(mixedRequest).execute()) {
+			assertEquals(mixedHandler.getMessageHeaders().getResponseStatusCode().code(), response.code());
+			assertEquals(mixedId, mixedHandler.lastReceivedRequest.index);
+		}
+	}
+
+	@Test
+	public void testJsonMultipart() throws Exception {
+		OkHttpClient client = new OkHttpClient();
+
+		int jsonId = RANDOM.nextInt();
+		Request jsonRequest = buildJsonRequest(jsonHandler.getMessageHeaders().getTargetRestEndpointURL(), jsonId);
+		try (Response response = client.newCall(jsonRequest).execute()) {
+			assertEquals(jsonHandler.getMessageHeaders().getResponseStatusCode().code(), response.code());
+			assertEquals(jsonId, jsonHandler.lastReceivedRequest.index);
+		}
+
+		Request fileRequest = buildFileRequest(jsonHandler.getMessageHeaders().getTargetRestEndpointURL());
+		try (Response response = client.newCall(fileRequest).execute()) {
+			// either because JSON payload is missing or FileUploads are outright forbidden
+			assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.code());
+		}
+
+		Request mixedRequest = buildMixedRequest(jsonHandler.getMessageHeaders().getTargetRestEndpointURL(), RANDOM.nextInt());
+		try (Response response = client.newCall(mixedRequest).execute()) {
+			// FileUploads are outright forbidden
+			assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.code());
+		}
+	}
+
+	@Test
+	public void testFileMultipart() throws Exception {
+		OkHttpClient client = new OkHttpClient();
+
+		Request jsonRequest = buildJsonRequest(fileHandler.getMessageHeaders().getTargetRestEndpointURL(), RANDOM.nextInt());
+		try (Response response = client.newCall(jsonRequest).execute()) {
+			// JSON payload did not match expected format
+			assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.code());
+		}
+
+		Request fileRequest = buildFileRequest(fileHandler.getMessageHeaders().getTargetRestEndpointURL());
+		try (Response response = client.newCall(fileRequest).execute()) {
+			assertEquals(fileHandler.getMessageHeaders().getResponseStatusCode().code(), response.code());
+		}
+
+		Request mixedRequest = buildMixedRequest(fileHandler.getMessageHeaders().getTargetRestEndpointURL(), RANDOM.nextInt());
+		try (Response response = client.newCall(mixedRequest).execute()) {
+			// JSON payload did not match expected format
+			assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.code());
+		}
+	}
```

Comment:
Can we add a test which assures that in case of a failure in `FileUploadHandler` the upload directory gets deleted?

Reply:
We should also add a test which verifies that the upload directory gets cleaned up after the request has been processed.

Reply:
it's a little tricky to test, since the handler responds to the request before cleaning the files. (since the response is sent within the try-with-resources statement)

Reply:
the underlying issue is that we don't have unit test for `AbstractHandler` and `AbstractRestHandler`.